service/dap/server.go

```diff
@@ -0,0 +1,420 @@
+		// dap.ReadProtocolMessage(), so we should never reach here.
+		panic(fmt.Sprintf("Unable to process %#v", request))
+	}
+}
+
+func (s *Server) send(message dap.Message) {
+	jsonmsg, _ := json.Marshal(message)
+	s.log.Debug("[-> to client]", string(jsonmsg))
+	dap.WriteProtocolMessage(s.conn, message)
+}
+
+func (s *Server) onInitializeRequest(request *dap.InitializeRequest) {
+	// TODO(polina): Respond with an error if debug session is in progress?
+	response := &dap.InitializeResponse{Response: *newResponse(request.Request)}
+	response.Body.SupportsConfigurationDoneRequest = true
+	// TODO(polina): support this to match vscode-go functionality
+	response.Body.SupportsSetVariable = false
+	s.send(response)
+}
+
+func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
+	// TODO(polina): Respond with an error if debug session is in progress?
+	program, ok := request.Arguments["program"]
+	if !ok || program == "" {
+		s.sendErrorResponse(request.Request,
+			3000, "Failed to launch",
+			"The program attribute is missing in debug configuration.")
+		return
+	}
+	s.config.ProcessArgs = []string{program.(string)}
+	s.config.WorkingDir = filepath.Dir(program.(string))
+	// TODO: support program args
+
+	stop, ok := request.Arguments["stopOnEntry"]
+	s.stopOnEntry = (ok && stop == true)
+
+	mode, ok := request.Arguments["mode"]
+	if !ok || mode == "" {
+		mode = "debug"
+	}
+	// TODO(polina): support "debug", "test" and "remote" modes
+	if mode != "exec" {
+		s.sendErrorResponse(request.Request,
+			3000, "Failed to launch",
+			fmt.Sprintf("Unsupported 'mode' value '%s' in debug configuration.", mode))
+		return
+	}
+
+	config := &debugger.Config{
+		WorkingDir:           s.config.WorkingDir,
+		AttachPid:            0,
+		CoreFile:             "",
+		Backend:              s.config.Backend,
+		Foreground:           s.config.Foreground,
+		DebugInfoDirectories: s.config.DebugInfoDirectories,
+		CheckGoVersion:       s.config.CheckGoVersion,
+	}
+	var err error
+	if s.debugger, err = debugger.New(config, s.config.ProcessArgs); err != nil {
+		s.sendErrorResponse(request.Request,
+			3000, "Failed to launch", err.Error())
+		return
+	}
+
+	// Notify the client that the debugger is ready to start "accepting"
+	// configuration requests for setting breakpoints, etc. The client
+	// will end the configuration sequence with 'configurationDone'.
+	s.send(&dap.InitializedEvent{Event: *newEvent("initialized")})
+	s.send(&dap.LaunchResponse{Response: *newResponse(request.Request)})
+}
+
+func (s *Server) onDisconnectRequest(request *dap.DisconnectRequest) {
+	s.send(&dap.DisconnectResponse{Response: *newResponse(request.Request)})
+	// TODO(polina): only halt if the program is running
+	_, err := s.debugger.Command(&api.DebuggerCommand{Name: api.Halt})
+	if err != nil {
+		s.log.Error(err)
+	}
+	kill := s.config.AttachPid == 0
+	err = s.debugger.Detach(kill)
+	if err != nil {
+		s.log.Error(err)
+	}
+	if s.config.DisconnectChan != nil {
```

Comment:
This needs careful documentation. The sequence "if not nil, close, then set to nil" is not atomic and will race if different goroutines execute it. Ideally, we'd have a single place where the channel is closed